A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,414 @@
+       if entry found then {
+         subchannel = subchannel from entry;
+         if subchannel is READY then {
+           return subchannel as the pick;
+         } else if subchannel is IDLE and idleSubchannel is none then {
+           idleSubchannel = subchannel;
+         } else if subchannel is CONNECTING {
+           foundConnecting = true;
+         }
+         equAddresses = equivalent addresses from entry;
+         newAddresses = subset of equAddresses not already present in equivalentAddressList;
+         add newAddresses to the end of equivalentAddressList;
+       }
+     }
+     if idleSubchannel is not none then {
+       trigger connection attempt on idleSubchannel;
+       return queue as pick result;
+     }
+     if foundConnecting then {
+       return queue as pick result;
+     }
+     return childPicker.pick();
+   }
+```
+
+In the above logic we prefer a subchannel in the `READY` state for a different
+equivalent address instead of waiting for the subchannel for the original
+`overrideHost` address to become `READY` (from one of `IDLE`, `CONNECTING`
+or `TRANSIENT_FAILURE` states) because we assume that the equivalent address
+is pointing to the same host thereby maintaining session affinity with the
+pick.
+
+If we do not find a `READY` subchannel, and find an `IDLE` one, we trigger a
+connection attempt on that subchannel and return queue as the pick result
+(i.e. the RPC stays buffered). If we instead find a `CONNECTING` subchannel,
+we just return queue as the pick result i.e. the RPC stays buffered until the
+connection attempt completes. If we do not find a subchannel, we just delegate
+to the child policy.
+
+Note that we unconditionally create the `override_host_experimental` policy
+as the child of `xds_cluster_impl_experimental` even if the feature is not
+configured (in which case, it will be a no-op). The
+`xds_cluster_resolver_experimental` policy's config will have
+a boolean called `override_host_enabled` and this config is passed down in
+the hierarchy. `cds_experimental` sets the boolean based on whether the
+[`common_lb_config.override_host_status`][or-host-status]
+field in the CDS update includes UNKNOWN or HEALTHY. If the field is unset,
+we treat it the same as if it was explicitly set to the default set.
+The diagram below shows how the configuration is passed down the hierarchy
+all the way from `cds_experimental` to `override_host_experimental`.
+
+If the `override_host_experimental` policy is disabled, it should still
+maintain the `overrideHostMap`, so that it has all of the necessary data
+to start working if it becomes enabled later, but it does not inject its
+own picker when disabled.
+
+
+One of the existing policies (`xds_wrr_locality_experimental`,
+or `ring_hash_experimental`) is created as the child policy of
+`override_host_experimental` as shown in the diagram below.
+
+![LB Policy Hierarchy Diagram](A55_graphics/lb-hierarchy.png)
```

Comment:
The diagram implies that the bool will be present in the configs for the `xds_cluster_resolver`, `priority`, and `xds_cluster_impl` policies.  I don't think we need that in any of them.

Reply:
fixed

Reply:
I just realized that what I described above was actually not entirely correct.  The `cds` policy does not actually generate everything below the `priority` policy directly, because in the case of an aggregate cluster, the config for features like outlier detection and circuit breaking are actually determined by the underlying clusters, not the aggregate cluster.  So the way we actually do this is to have these settings added to the `xds_cluster_resolver` policy's config, which contains a separate `DiscoveryMechanism` for each underlying cluster.  I think we'll need to use that same approach for the `override_host_status` field.

In other words, what we need here is the following:
- The `cds` policy will copy the `override_host_status` value it gets from the `XdsClient` to the appropriate `DiscoveryMechanism` entry in the config of the `xds_cluster_resolver` policy.  (Note: This will require adding a new field to the `xds_cluster_resolver` LB policy config.)
- The `xds_cluster_resolver` policy will copy the `override_host_status` value from its config into the config for the `xds_override_host` policy, which it will generate (just like it already does for the `outlier_detection` policy).

Reply:
removed the red line in the diagram and fixed the diagram and description to include `DiscoveryMechanism` of `xds_cluster_resolver`